File: cli-node/src/commands/scan.ts
import fs from "node:fs";
import path from "node:path";

import chalk from "chalk";
import yaml from "js-yaml";

/**
 * scopeform scan — free, local shadow-agent / raw-credential detector.
 *
 * Runs fully locally: nothing is sent anywhere and no login is required.
 */

export type Finding = {
  file: string;
  line: number;
  kind: string;
  service: string | null;
  risk: "high" | "medium";
  detail: string;
};

// Provider-accurate secret patterns; each maps to the service we can scope.
const SECRET_PATTERNS: Array<[string, string, "high" | "medium", RegExp]> = [
  ["Anthropic API key", "anthropic", "high", /sk-ant-[A-Za-z0-9_-]{20,}/],
  ["OpenAI API key", "openai", "high", /sk-(?!ant-)(?:proj-)?[A-Za-z0-9_-]{20,}/],
  ["GitHub token", "github", "high", /(?:gh[pousr]_[A-Za-z0-9]{36,}|github_pat_[A-Za-z0-9_]{22,})/],
  ["Slack token", "slack", "high", /xox[baprs]-[A-Za-z0-9-]{10,}/],
  ["AWS access key", "aws", "high", /AKIA[0-9A-Z]{16}/],
  ["Stripe live key", "stripe", "high", /[sr]k_live_[A-Za-z0-9]{24,}/],
  ["Google API key", "google", "medium", /AIza[0-9A-Za-z_-]{35}/],
];

const ENV_ASSIGNMENT = /^\s*(?<name>[A-Z0-9_]*(?:API_KEY|APIKEY|SECRET|TOKEN|PASSWORD)[A-Z0-9_]*)\s*=\s*(?<value>[^\s#]{12,})/;

const PLACEHOLDER_HINTS = ["placeholder", "your", "xxx", "changeme", "example", "dummy", "<", "${", "replace", "todo"];

const WORKFLOW_SECRET = /\$\{\{\s*secrets\.(?<name>[A-Za-z0-9_]+)\s*\}\}/g;

const SOURCE_SUFFIXES = new Set([".py", ".js", ".ts", ".jsx", ".tsx", ".mjs", ".cjs"]);
const CONFIG_SUFFIXES = new Set([".json", ".yml", ".yaml", ".toml", ".cfg", ".ini"]);

const SKIP_DIRS = new Set([
  ".git", "node_modules", "dist", "build", "out", ".next", "coverage",
  ".venv", "venv", "env", "__pycache__", ".mypy_cache", ".pytest_cache",
  ".ruff_cache", "target", "vendor", ".turbo",
]);

const MAX_FILE_BYTES = 1_000_000;

const DEFAULT_ACTIONS: Record<string, string[]> = {
  openai: ["chat.completions"],
  anthropic: ["messages"],
  github: ["repos.read"],
};

function redact(value: string): string {
  return value.length > 8 ? `${value.slice(0, 8)}…` : "…";
}

function looksPlaceholder(value: string): boolean {
  const lowered = value.toLowerCase();
  return PLACEHOLDER_HINTS.some((hint) => lowered.includes(hint));
}

function* iterFiles(root: string): Generator<string> {
  const stack = [root];
  while (stack.length > 0) {
    const dir = stack.pop() as string;
    let entries: fs.Dirent[];
    try {
      entries = fs.readdirSync(dir, { withFileTypes: true });
    } catch {
      continue;
    }
    for (const entry of entries) {
      const full = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        if (!SKIP_DIRS.has(entry.name)) {
          stack.push(full);
        }
      } else if (entry.isFile()) {
        try {
          if (fs.statSync(full).size <= MAX_FILE_BYTES) {
            yield full;
          }
        } catch {
          // unreadable — skip
        }
      }
    }
  }
}

function scanLineForSecrets(rel: string, lineNo: number, line: string): Finding[] {
  const findings: Finding[] = [];
  for (const [kind, service, risk, pattern] of SECRET_PATTERNS) {
    const match = pattern.exec(line);
    if (match && !looksPlaceholder(match[0])) {
      findings.push({
        file: rel,
        line: lineNo,
        kind,
        service,
        risk,
        detail: `${kind} (${redact(match[0])}) — unscoped, unrevocable, unaudited`,
      });
    }
  }
  return findings;
}

function readLines(file: string): string[] {
  try {
    return fs.readFileSync(file, "utf8").split("\n");
  } catch {
    return [];
  }
}

function scanEnvFile(file: string, rel: string): Finding[] {
  const findings: Finding[] = [];
  readLines(file).forEach((line, idx) => {
    const specific = scanLineForSecrets(rel, idx + 1, line);
    if (specific.length > 0) {
      findings.push(...specific);
      return;
    }
    const match = ENV_ASSIGNMENT.exec(line);
    const name = match?.groups?.name;
    const value = match?.groups?.value;
    if (name && value && !looksPlaceholder(value) && name !== "SCOPEFORM_TOKEN") {
      findings.push({
        file: rel,
        line: idx + 1,
        kind: "Credential in .env",
        service: null,
        risk: "medium",
        detail: `${name}=${redact(value)} — long-lived credential in plain text`,
      });
    }
  });
  return findings;
}

function scanTextFile(file: string, rel: string): Finding[] {
  const findings: Finding[] = [];
  readLines(file).forEach((line, idx) => {
    findings.push(...scanLineForSecrets(rel, idx + 1, line));
  });
  return findings;
}

function scanWorkflow(file: string, rel: string): Finding[] {
  const findings: Finding[] = [];
  readLines(file).forEach((line, idx) => {
    for (const match of line.matchAll(WORKFLOW_SECRET)) {
      const name = match.groups?.name;
      if (name && name !== "SCOPEFORM_TOKEN" && name !== "GITHUB_TOKEN") {
        findings.push({
          file: rel,
          line: idx + 1,
          kind: "Secret in CI workflow",
          service: null,
          risk: "medium",
          detail: `secrets.${name} passed directly to a workflow step — prefer a fresh scoped token per run`,
        });
      }
    }
  });
  return findings;
}

export function scanDirectory(root: string): Finding[] {
  const findings: Finding[] = [];
  for (const file of iterFiles(root)) {
    const rel = path.relative(root, file).split(path.sep).join("/");
    const base = path.basename(file);
    const suffix = path.extname(file);

    if (base === "scopeform.yml") {
      continue;
    }
    if (base.startsWith(".env")) {
      findings.push(...scanEnvFile(file, rel));
    } else if (rel.includes(".github/workflows/") && (suffix === ".yml" || suffix === ".yaml")) {
      findings.push(...scanWorkflow(file, rel));
    } else if (SOURCE_SUFFIXES.has(suffix) || CONFIG_SUFFIXES.has(suffix)) {
      findings.push(...scanTextFile(file, rel));
    }
  }
  return findings;
}

export function buildSuggestedConfig(findings: Finding[]): Record<string, unknown> | null {
  const services = [...new Set(findings.map((f) => f.service).filter((s): s is string => !!s && s in DEFAULT_ACTIONS))].sort();
  if (services.length === 0) {
    return null;
  }
  return {
    identity: { name: "my-agent", owner: "you@example.com", environment: "production" },
    ttl: "24h",
    scopes: services.map((service) => ({ service, actions: DEFAULT_ACTIONS[service] })),
  };
}

export async function scanCommand(target: string, options: { json?: string }): Promise<void> {
  const root = path.resolve(target);
  if (!fs.existsSync(root) || !fs.statSync(root).isDirectory()) {
    console.error(chalk.bold.red(`Not a directory: ${root}`));
    process.exitCode = 2;
    return;
  }

  const findings = scanDirectory(root);

  if (options.json) {
    fs.writeFileSync(options.json, JSON.stringify({ root, findings }, null, 2), "utf8");
  }

  if (findings.length === 0) {
    console.log(chalk.green("✓ No raw agent credentials found."));
    console.log("Nothing to fix — or your keys are already behind scoped tokens.");
    return;
  }

  console.log(chalk.bold(`Scopeform Scan — ${findings.length} finding(s)`));
  const sorted = [...findings].sort((a, b) =>
    a.risk === b.risk ? a.file.localeCompare(b.file) || a.line - b.line : a.risk === "high" ? -1 : 1,
  );
  for (const finding of sorted) {
    const risk = finding.risk === "high" ? chalk.red("HIGH") : chalk.yellow("MED ");
    console.log(`${risk}  ${finding.file}:${finding.line}  ${finding.detail}`);
  }

  const suggestion = buildSuggestedConfig(findings);
  if (suggestion) {
    console.log(chalk.bold("\nSuggested scopeform.yml — scoped, short-lived, revocable:"));
    console.log(yaml.dump(suggestion, { sortKeys: false }).trimEnd());
    console.log(`\nRun ${chalk.cyan("scopeform init")} then ${chalk.cyan("scopeform deploy")} to replace these keys with a scoped token.`);
  }

  if (options.json) {
    console.log(`\nReport written to ${chalk.cyan(options.json)}`);
  }

  process.exitCode = 1;
}

File: cli-py/scopeform/commands/scan.py
from __future__ import annotations

"""scopeform scan — free, local shadow-agent / raw-credential detector.

Runs fully locally: nothing is sent anywhere and no login is required.
Finds raw provider API keys in .env files, source code, and config files,
plus GitHub Actions workflows that hand secrets straight to scripts, then
suggests the scopeform.yml that would replace them with scoped tokens.
"""

import json
import re
from dataclasses import asdict, dataclass
from pathlib import Path

import typer
import yaml
from rich.console import Console
from rich.table import Table

console = Console()

# ── What we look for ──────────────────────────────────────────────────────────

# Provider-accurate secret patterns. Each maps to the service we can scope.
SECRET_PATTERNS: list[tuple[str, str, str, str]] = [
    # (finding label, service for scopeform.yml, risk, regex)
    ("Anthropic API key", "anthropic", "high", r"sk-ant-[A-Za-z0-9_\-]{20,}"),
    ("OpenAI API key", "openai", "high", r"sk-(?!ant-)(?:proj-)?[A-Za-z0-9_\-]{20,}"),
    ("GitHub token", "github", "high", r"(?:gh[pousr]_[A-Za-z0-9]{36,}|github_pat_[A-Za-z0-9_]{22,})"),
    ("Slack token", "slack", "high", r"xox[baprs]-[A-Za-z0-9\-]{10,}"),
    ("AWS access key", "aws", "high", r"AKIA[0-9A-Z]{16}"),
    ("Stripe live key", "stripe", "high", r"[sr]k_live_[A-Za-z0-9]{24,}"),
    ("Google API key", "google", "medium", r"AIza[0-9A-Za-z_\-]{35}"),
]

# Generic KEY=value assignments in .env-style files.
ENV_ASSIGNMENT = re.compile(
    r"^\s*(?P<name>[A-Z0-9_]*(?:API_KEY|APIKEY|SECRET|TOKEN|PASSWORD)[A-Z0-9_]*)\s*=\s*(?P<value>[^\s#]{12,})",
)

# Values that are clearly placeholders, not live credentials.
PLACEHOLDER_HINTS = ("placeholder", "your", "xxx", "changeme", "example", "dummy", "<", "${", "REPLACE", "todo")

# GitHub Actions: secrets handed straight to run steps.
WORKFLOW_SECRET = re.compile(r"\$\{\{\s*secrets\.(?P<name>[A-Za-z0-9_]+)\s*\}\}")

SOURCE_SUFFIXES = {".py", ".js", ".ts", ".jsx", ".tsx", ".mjs", ".cjs"}
CONFIG_SUFFIXES = {".json", ".yml", ".yaml", ".toml", ".cfg", ".ini"}

SKIP_DIRS = {
    ".git", "node_modules", "dist", "build", "out", ".next", "coverage",
    ".venv", "venv", "env", "__pycache__", ".mypy_cache", ".pytest_cache",
    ".ruff_cache", "target", "vendor", ".turbo",
}

MAX_FILE_BYTES = 1_000_000  # skip anything bigger — not a config/source file


@dataclass
class Finding:
    file: str
    line: int
    kind: str          # e.g. "OpenAI API key"
    service: str | None  # scopeform service name when known
    risk: str          # high | medium
    detail: str        # human explanation with the REDACTED value


def _redact(value: str) -> str:
    return f"{value[:8]}…" if len(value) > 8 else "…"


def _looks_placeholder(value: str) -> bool:
    lowered = value.lower()
    return any(hint.lower() in lowered for hint in PLACEHOLDER_HINTS)


def _iter_files(root: Path):
    for path in root.rglob("*"):
        if path.is_dir():
            continue
        if any(part in SKIP_DIRS for part in path.parts):
            continue
        try:
            if path.stat().st_size > MAX_FILE_BYTES:
                continue
        except OSError:
            continue
        yield path


def _scan_line_for_secrets(rel: str, line_no: int, line: str) -> list[Finding]:
    findings: list[Finding] = []
    for kind, service, risk, pattern in SECRET_PATTERNS:
        for match in re.finditer(pattern, line):
            value = match.group(0)
            if _looks_placeholder(value):
                continue
            findings.append(
                Finding(
                    file=rel,
                    line=line_no,
                    kind=kind,
                    service=service,
                    risk=risk,
                    detail=f"{kind} ({_redact(value)}) — unscoped, unrevocable, unaudited",
                )
            )
            break  # one finding per pattern per line is enough
    return findings


def _scan_env_file(path: Path, rel: str) -> list[Finding]:
    findings: list[Finding] = []
    try:
        lines = path.read_text(encoding="utf-8", errors="ignore").splitlines()
    except OSError:
        return findings

    for idx, line in enumerate(lines, start=1):
        specific = _scan_line_for_secrets(rel, idx, line)
        if specific:
            findings.extend(specific)
            continue
        match = ENV_ASSIGNMENT.match(line)
        if match and not _looks_placeholder(match.group("value")):
            name = match.group("name")
            if name == "SCOPEFORM_TOKEN":
                continue  # already scoped — that's the fix, not a finding
            findings.append(
                Finding(
                    file=rel,
                    line=idx,
                    kind="Credential in .env",
                    service=None,
                    risk="medium",
                    detail=f"{name}={_redact(match.group('value'))} — long-lived credential in plain text",
                )
            )
    return findings


def _scan_text_file(path: Path, rel: str) -> list[Finding]:
    findings: list[Finding] = []
    try:
        lines = path.read_text(encoding="utf-8", errors="ignore").splitlines()
    except OSError:
        return findings
    for idx, line in enumerate(lines, start=1):
        findings.extend(_scan_line_for_secrets(rel, idx, line))
    return findings


def _scan_workflow(path: Path, rel: str) -> list[Finding]:
    findings: list[Finding] = []
    try:
        lines = path.read_text(encoding="utf-8", errors="ignore").splitlines()
    except OSError:
        return findings
    for idx, line in enumerate(lines, start=1):
        for match in WORKFLOW_SECRET.finditer(line):
            name = match.group("name")
            if name in ("SCOPEFORM_TOKEN", "GITHUB_TOKEN"):
                continue  # scoped/native tokens are fine
            findings.append(
                Finding(
                    file=rel,
                    line=idx,
                    kind="Secret in CI workflow",
                    service=None,
                    risk="medium",
                    detail=f"secrets.{name} passed directly to a workflow step — prefer a fresh scoped token per run",
                )
            )
    return findings


def scan_directory(root: Path) -> list[Finding]:
    findings: list[Finding] = []
    for path in _iter_files(root):
        rel = path.relative_to(root).as_posix()
        name = path.name

        if name == "scopeform.yml":
            continue
        if name.startswith(".env"):
            findings.extend(_scan_env_file(path, rel))
        elif ".github/workflows/" in rel and path.suffix in (".yml", ".yaml"):
            findings.extend(_scan_workflow(path, rel))
        elif path.suffix in SOURCE_SUFFIXES or path.suffix in CONFIG_SUFFIXES:
            findings.extend(_scan_text_file(path, rel))
    return findings


# ── Suggested fix ─────────────────────────────────────────────────────────────

DEFAULT_ACTIONS = {
    "openai": ["chat.completions"],
    "anthropic": ["messages"],
    "github": ["repos.read"],
}


def build_suggested_config(findings: list[Finding]) -> dict | None:
    services = sorted({f.service for f in findings if f.service and f.service in DEFAULT_ACTIONS})
    if not services:
        return None
    return {
        "identity": {
            "name": "my-agent",
            "owner": "you@example.com",
            "environment": "production",
        },
        "ttl": "24h",
        "scopes": [{"service": svc, "actions": DEFAULT_ACTIONS[svc]} for svc in services],
    }


# ── Command ───────────────────────────────────────────────────────────────────

def scan_command(
    path: Path = typer.Argument(Path("."), help="Directory to scan (default: current directory)."),
    json_out: Path | None = typer.Option(None, "--json", help="Also write the findings report to a JSON file."),
) -> None:
    """Scan for raw agent credentials. Fully local — nothing leaves this machine."""
    root = path.resolve()
    if not root.is_dir():
        console.print(f"[bold red]Not a directory: {root}[/bold red]")
        raise typer.Exit(code=2)

    with console.status(f"Scanning {root}..."):
        findings = scan_directory(root)

    if json_out is not None:
        json_out.write_text(
            json.dumps({"root": str(root), "findings": [asdict(f) for f in findings]}, indent=2),
            encoding="utf-8",
        )

    if not findings:
        console.print("[green]✓ No raw agent credentials found.[/green]")
        console.print("Nothing to fix — or your keys are already behind scoped tokens.")
        raise typer.Exit(code=0)

    table = Table(title=f"Scopeform Scan — {len(findings)} finding(s)")
    table.add_column("Risk", style="bold")
    table.add_column("Location")
    table.add_column("Finding")
    for f in sorted(findings, key=lambda f: (f.risk != "high", f.file, f.line)):
        risk_style = "red" if f.risk == "high" else "yellow"
        table.add_row(f"[{risk_style}]{f.risk.upper()}[/{risk_style}]", f"{f.file}:{f.line}", f.detail)
    console.print(table)

    suggestion = build_suggested_config(findings)
    if suggestion is not None:
        console.print("\n[bold]Suggested scopeform.yml[/bold] — scoped, short-lived, revocable:")
        console.print(yaml.safe_dump(suggestion, sort_keys=False).rstrip())
        console.print("\nRun [cyan]scopeform init[/cyan] then [cyan]scopeform deploy[/cyan] to replace these keys with a scoped token.")

    if json_out is not None:
        console.print(f"\nReport written to [cyan]{json_out}[/cyan]")

    raise typer.Exit(code=1)
